### PR TITLE
Made several changes to min-cut partitioner that allow it to recompute     more things

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122689
* #122581
* __->__ #121692
* #122688
* #122686

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Perf results
image

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -379,31 +379,6 @@ def pointwise_ops():
 
     return ops
 
-def get_depth(node, depth_map):
```
Comment (Chillee) [inline torch/_functorch/partitioners.py:382]: python unfortunately has a very low recursion limit (1000), so rewrote into a non-recursive algorithm.

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -733,10 +722,10 @@ def classify_nodes(joint_module):
...
     if compiler == "inductor":
-        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, aten.var_mean, aten.isinf, aten.any, aten.full, aten.as_strided, aten.zeros, aten.argmax, aten.maximum]  # noqa: E501,B950
-        view_ops += [aten.view, aten.slice, aten.permute, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided]
+        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, aten.var_mean, aten.isinf, aten.any, aten.full, aten.as_strided, aten.zeros, aten.argmax, aten.maximum, prims.iota]  # noqa: E501,B950
```
Comment (eellison) [inline torch/_functorch/partitioners.py:725]: Can we use tags here ? It's not surprising that are missing operators.

File: torch/_functorch/partitioners.py [hunk 3]
```diff
@@ -733,10 +722,10 @@ def classify_nodes(joint_module):
...
+        view_ops += [aten.view, aten.slice, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided, aten.permute]
```
Comment (eellison) [inline torch/_functorch/partitioners.py:726]: Is there a better way of maintaining this list too ?

File: torch/_functorch/partitioners.py (hunk 2 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:725]: I don't think we have a view tag right?

Comment (peterbell10) [reply torch/_functorch/partitioners.py:725]: If you had a function `is_recomputable(node)` then it could check 
```python
get_node_storage(node) == get_node_storage(node.args[0])
```

File: torch/_functorch/partitioners.py (hunk 3 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:726]: I'll add a view tag in a follow-up PR

File: torch/_inductor/triton_heuristics.py [hunk 4]
```diff
@@ -339,6 +339,7 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
                 "num_warps": compile_meta["num_warps"],
                 "num_stages": compile_meta["num_stages"],
                 "debug": compile_meta["debug"],
+                "enable_fp_fusion": False,
```
Comment (shunting314) [inline torch/_inductor/triton_heuristics.py:342]: This is to disable FMA?

I'm wondering if this should be in a separate PR even though it's just a single line. I know that you recently posted about an numerical issue caused by FMA. But on the other hand, theoretically speaking disabling FMA may have perf hit due to more computation instructions. Do we observe any perf impact?

Comment (Chillee) [reply torch/_inductor/triton_heuristics.py:342]: Yeah I'll put it in a separate PR - @peterbell10 is working on that.

File: torch/_functorch/config.py [hunk 5]
```diff
@@ -32,11 +32,32 @@
 cse = True
 
 # Restricts the amount of computation AOTAutograd can do.
-max_dist_from_bw = 3
+max_dist_from_bw = 1000
```
Comment (eellison) [inline torch/_functorch/config.py:35]: where is this number coming from ? if it's this large, isnt even worth having ?

File: torch/_functorch/partitioners.py [hunk 6]
```diff
@@ -452,7 +426,11 @@ def reordering_to_mimic_autograd_engine(gm):
     # Populate depth for the nodes. Depth is the distance from the inputs.
     depths = {}
     output_node = next(node for node in gm.graph.nodes if node.op == "output")
-    get_depth(output_node, depths)
+    for node in gm.graph.nodes:
+        if node.op == 'placeholder':
+            depths[node] = 0
+        else:
+            depths[node] = max([depths[arg] for arg in node.all_input_nodes if isinstance(arg, torch.fx.Node)], default=0)
```
Comment (eellison) [inline torch/_functorch/partitioners.py:433]: `for arg in node.all_input_nodes if isinstance(arg, torch.fx.Node)`

dont think you need the isinstance here with `node.all_input_nodes`

File: torch/_functorch/partitioners.py [hunk 7]
```diff
@@ -474,7 +452,8 @@ def insert_node_in_graph(node):
             if order[user] < minimum_order:
                 minimum_order = order[user]
                 first_node_in_bwd = user
-    assert first_node_in_bwd is not None
+    if first_node_in_bwd is None:
+        return gm
```
Comment (eellison) [inline torch/_functorch/partitioners.py:456]: why is this ?

File: torch/_functorch/partitioners.py [hunk 8]
```diff
@@ -718,6 +697,19 @@ def classify_nodes(joint_module):
     if len(required_bw_nodes) == 0:
         return default_partition(joint_module, _joint_inputs, num_fwd_outputs=num_fwd_outputs)
 
+    def is_fusible(a, b):
+        # We can perform "memory fusion" into a cat, but cat cannot be a
+        # producer to a fusion
+        if get_aten_target(b) == aten.cat:
+            return True
```
Comment (eellison) [inline torch/_functorch/partitioners.py:704]: should this be ?
```
if get_aten_target(b) == aten.cat and get_aten_target(a): 
	return False
```

File: torch/_functorch/partitioners.py (hunk 3 above)
Comment (eellison) [reply torch/_functorch/partitioners.py:729]: that or @peterbell10 suggestion

File: torch/_functorch/partitioners.py [hunk 9]
```diff
@@ -749,11 +741,10 @@ def classify_nodes(joint_module):
         method_to_operator(m)
         for m in magic_methods
     ]
-
     recomputable_ops = set(recomputable_ops) if recomputable_ops is not None else set(default_recomputable_ops)
 
     random_ops = [aten.native_dropout, aten.rand_like, aten.randn_like]
-    compute_intensive_ops = [aten.mm, aten.convolution, aten.convolution_backward, aten.bmm, aten.addmm, aten.upsample_bilinear2d, aten._softmax, aten._softmax_backward_data, aten.native_layer_norm, aten.native_layer_norm_backward, aten.native_batch_norm, aten.native_batch_norm_backward, aten._native_batch_norm_legit, aten._batch_norm_with_update, aten.batch_norm_backward]  # noqa: E501,B950
+    compute_intensive_ops = [aten.mm, aten.convolution, aten.convolution_backward, aten.bmm, aten.addmm, aten._scaled_dot_product_flash_attention, aten._scaled_dot_product_efficient_attention, aten.upsample_bilinear2d]  # noqa: E501,B950
```
Comment (eellison) [inline torch/_functorch/partitioners.py:747]: batchnorm/softmax/layernorm were able to be removed bc of reduction changes ? how did you decide upsample_bilinear2. 

i think you might want max_pool2d_backward as well

File: torch/_functorch/partitioners.py [hunk 10]
```diff
@@ -766,64 +757,76 @@ def classify_nodes(joint_module):
...
+        # backwards pass is "free". However, if a node must be materialized
+        # in the backwards pass, then recomputing it is never free.
+        if is_materialized_backwards(node) and BAN_IF_MATERIALIZED_BACKWARDS:
+            log.info("materialized backwards: %s %s", node, tuple(node.users))
+            return True
+
+        # Arbitrary hack that sometimes seems to help things. The above
+        # modification appears to have made this heuristic a lot less critical
+        # for performance.
+        if not graph_has_recomputable_ops:
```
Comment (eellison) [inline torch/_functorch/partitioners.py:815]: which heuristic specifically ? `is_materialized_backwards` ?

File: torch/_functorch/partitioners.py [hunk 11]
```diff
@@ -837,15 +840,27 @@ def get_node_weight(node) -> int:
...
+    def ban_recomputation_if_allowed(node):
+        # NB: "recompute" == 0 means it's not banned from recomputing
+        if node.meta.get("recompute", 0) > 0:
+            return False
+        banned_nodes.add(node)
+        # A node will only ever be recomputed if there is a path from an
+        # ancestor of this node to the backwards path through this node that
+        # doesn't go through any saved value. If this node is saved, then that
+        # condition is not possible.
+        nx_graph.add_edge("source", node.name + "_in", capacity=math.inf)
```
Comment (eellison) [inline torch/_functorch/partitioners.py:860]: worth factoring out the inf banning of recomputation into helper ?

File: torch/_functorch/partitioners.py [hunk 12]
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
         for user in node.users:
             nx_graph.add_edge(node.name + "_out", user.name + "_in", capacity=math.inf)
 
+    # todo(chilli): This is the most questionable of the 3 heuristics for banning recompute.
+    # I would look more into it but I'm out of time for now
```
Comment (eellison) [inline torch/_functorch/partitioners.py:907]: you can prob leave this sentence out haha. that's what are all todos are for lol

File: torch/_functorch/partitioners.py [hunk 13]
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
...
+
+    # Thus, we shouldn't aim to recompute far downstream nodes that depend on
+    # this node. That intuition of "far downstream" is captured by whether
+    # there's an unfusible op along the chain somewhere
+
+    # It could probably be improved by properly analyzing what's going on in the
+    # backwards pass instead of only relying on whether it's unfusible in the
+    # forwards.
+
+    def find_first_unfusible(start_nodes, max_range):
```
Comment (eellison) [inline torch/_functorch/partitioners.py:924]: nit: add type annotation, docstring ?

File: torch/_functorch/partitioners.py [hunk 14]
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
...
+        sorted_nodes = []
+        for n in start_nodes:
```
Comment (eellison) [inline torch/_functorch/partitioners.py:926]: still need to review rest of this file, will do tomorrow

**Review by eellison — COMMENTED** [review]
part way through review

File: torch/_functorch/config.py (hunk 5 above)
Comment (Chillee) [reply torch/_functorch/config.py:35]: Yeah, it's basically removed now - keeping it for now as a knob in case somebody finds it necessary in the short term.

File: benchmarks/dynamo/timm_models.py [hunk 15]
```diff
@@ -262,12 +263,21 @@ def load_model(
...
-        std_dev = torch.std(input_tensor)
-        example_inputs = (input_tensor - mean) / std_dev
+        from urllib.request import urlopen
+        from PIL import Image
+
+        img = Image.open(urlopen(
+            'https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/beignets-task-guide.png'
+        ))
+        transforms = create_transform(**data_config, is_training=False)
+        example_inputs = transforms(img).to(device=device).half().unsqueeze(0).repeat(batch_size, *([1]*len(input_size)))
```
[the remark this answers is not in the document]
Comment (Chillee) [reply benchmarks/dynamo/timm_models.py]: Ah I've since removed these changes.

File: torch/_dynamo/utils.py [hunk 16]
```diff
@@ -364,12 +364,12 @@ def setup_compile_debug():
     compile_debug = os.environ.get("TORCH_COMPILE_DEBUG", "0") == "1"
 
     if compile_debug:
-        torch._logging.set_logs(
-            dynamo=logging.DEBUG,
-            aot=logging.DEBUG,
-            inductor=logging.DEBUG,
-            output_code=True,  # this is off by default
-        )
+        # torch._logging.set_logs(
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_dynamo/utils.py]: Was gonna disable this, but will push to a later PR.

File: torch/_functorch/partitioners.py (hunk 7 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:456]: Fixes a bug in existing impl - if gradInp does not depend on gradOut then this will not detect any nodes in the bacwkards.

Added a comment.

File: torch/_functorch/partitioners.py (hunk 8 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:704]: The idea is that you can always "fuse" ops into cat, so recomputing an op that feeds into cat will always be free. But this was here before - this is just code movement.

File: torch/_functorch/partitioners.py (hunk 9 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:747]: They were removed since we never see them nowadays at this point.

This list is not used unless aggressive recompute is turned on. In that case, I think it might make more sense to not recompute things like upsampling. This is more preparation for later changes for more user control for recomputation.

File: torch/_functorch/partitioners.py (hunk 10 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:815]: Yeah. Added a comment noting its obsolescence.

File: torch/_functorch/partitioners.py (hunk 11 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:860]: This is the helper haha.

File: torch/_functorch/partitioners.py [hunk 17]
```diff
@@ -921,10 +1022,14 @@ def get_node_weight(node) -> int:
             fw_module, bw_module = functionalize_rng_ops(
                 joint_module, fw_module, bw_module, len(saved_sym_nodes)
             )
-        bw_module = reordering_to_mimic_autograd_engine(bw_module)
+    bw_module = reordering_to_mimic_autograd_engine(bw_module)
```
Comment (Chillee) [inline torch/_functorch/partitioners.py:1025]: This is deliberate btw.

File: torch/_functorch/partitioners.py [hunk 18]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
+        visited = set()
+        for start_node in joint_graph.nodes:
+            if start_node not in required_fw_nodes:
+                continue
+            fusible = [(start_node.fw_order, start_node)]
+            start_pos = start_node.fw_order
```
Comment (shunting314) [inline torch/_functorch/partitioners.py:976]: not used

File: torch/_functorch/partitioners.py [hunk 19]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+            start_order = start_node.fw_order
+            while len(fusible) > 0:
+                _, cur = heapq.heappop(fusible)
+                if cur in visited:
+                    continue
+                visited.add(cur)
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
```
Comment (shunting314) [inline torch/_functorch/partitioners.py:983]: 100 here does not necessarily means the chains contains 100 nodes right? Most likely the length of the chain is much less than 100.

File: torch/_functorch/partitioners.py [hunk 20]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+
+
+    # This heuristic is fairly straightforward. The idea is that although it is
+    # cheap to recompute bandwidth-bound ops, we don't want to end up in a situation
+    # where we have a long chain of pointwise ops from the beginning to the end
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
```
Comment (shunting314) [inline torch/_functorch/partitioners.py:970]: A general question about this heuristic, let's say I have a long chain of nodes that I can fuse: n1, n2, ...., n100. Sounds like we should ban the recomputation for n100.

The chain is long, but if we save say n98, then when we recompute n100 in backward pass, we only need recompute n99 and n100. The recomputation cost is not high even though the chain of fusable nodes is long (100 nodes in this case). Then ban recomputation for n100 seems to be not necessary for this specific case? (The example pick an unrealisticaly long chain for discussion purpose)

File: torch/_functorch/partitioners.py (hunk 19 above)
Comment (shunting314) [reply torch/_functorch/partitioners.py:983]: I think it's possible to accurately maintain the chain length similar to how `get_depth` works. Not sure if that's necessary though

File: torch/_functorch/partitioners.py [hunk 21]
```diff
@@ -379,31 +382,6 @@ def pointwise_ops():
 
     return ops
 
-def get_depth(node, depth_map):
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:382]: Yeah. In this case it was specifically applied to the backwards pass.

File: torch/_functorch/partitioners.py (hunk 19 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:983]: > 100 here does not necessarily means the chains contains 100 nodes right? 

Yeah, I don't think this heuristic is that important actually.

File: torch/_functorch/partitioners.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:382]: Yeah. This is also only applied to the backwards graph.

File: torch/_functorch/partitioners.py (hunk 20 above)
Comment (Chillee) [reply torch/_functorch/partitioners.py:970]: If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

IMO, 2 things:
1. This is kind of a degenerate case, and so I think the precise performance here doesn't matter too much as long as we're not falling into the degenerate case.
2. We do have a heuristic there where we only save it once it narrows down to a "bottleneck".

File: torch/_functorch/partitioners.py [hunk 22]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
+        visited = set()
+        for start_node in joint_graph.nodes:
+            if start_node not in required_fw_nodes:
+                continue
+            fusible = [(start_node.fw_order, start_node)]
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:975]: I don't think Python has another heap data structure 🤔 https://stackoverflow.com/questions/12749622/understanding-how-to-create-a-heap-in-python

File: torch/_functorch/partitioners.py [hunk 23]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+                        )
+                        ban_recomputation_if_allowed(user)
+
+
+    # This heuristic is fairly straightforward. The idea is that although it is
+    # cheap to recompute bandwidth-bound ops, we don't want to end up in a situation
+    # where we have a long chain of pointwise ops from the beginning to the end
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:968]: image

File: torch/_functorch/partitioners.py [hunk 24]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+                continue
+            fusible = [(start_node.fw_order, start_node)]
+            start_order = start_node.fw_order
+            while len(fusible) > 0:
+                _, cur = heapq.heappop(fusible)
+                if cur in visited:
+                    continue
+                visited.add(cur)
+                # 100 is arbitrary choice to try and prevent degenerate cases
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:983]: I think right now all the choices are just booleans, so I hesitate on making it an integer.

File: torch/_functorch/partitioners.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:970]: We could explore it in follow-up commits but I don't want to retest/isolate the affect of this change.

File: torch/_functorch/partitioners.py [hunk 25]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+                first_unfusible_use = find_first_unfusible(fw_users, max(orders))
+                for user in tuple(used_node.users):
+                    if user in required_fw_nodes and user.fw_order > first_unfusible_use and is_fusible(used_node, user):
+                        if user in banned_nodes:
+                            continue
+                        log.info(
+                            "used above/below fusible %s:(%s) -> %s -> %s:(%s)",
+                            used_node, used_node.fw_order, first_unfusible_use, user, user.fw_order
+                        )
+                        ban_recomputation_if_allowed(user)
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:960]: The idea is that if we have something like
```
a: Tensor
b = cos(a)
c = mm(a)
...
d = add(a, c)
```

`b` and `d` cannot get fused together, and so if we choose to recompute `d` we will be reading from `a` multiple times.

It's not a perfect heuristic, but it's what I have now :)

File: torch/_functorch/partitioners.py [hunk 26]
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
...
+                    continue
+                visited.add(cur)
+                # 100 is arbitrary choice to try and prevent degenerate cases
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
+                    log.info("too long %s %s %s %s", cur, start_node, cur.fw_order, start_node.fw_order)
+                    ban_recomputation_if_allowed(cur)
+                    break
+
+                for user in cur.users:
+                    if user in required_fw_nodes and is_fusible(cur, user) and user not in banned_nodes:
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_functorch/partitioners.py:989]: It might force a materialization yeah, but since this is only nodes in the forwards pass that's not something that "affects" recomputation.

**Comment by Chillee** [comment]
@eellison For the most part, none of the heuristics are explicitly trying to work around Inductor limitations AFAIK. Some of the heuristics definitely aren't perfect, but that's why they're heuristics :)

File: torch/_functorch/partitioners.py (hunk 20 above)
Comment (shunting314) [reply torch/_functorch/partitioners.py:970]: > If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

Yes. Due to saving n98, recomputing n100 is not that expensive even though the fusible chain is very long.

**Review by shunting314 — APPROVED** [review]

**Review by eellison — APPROVED** [review]

File: torch/_functorch/partitioners.py [hunk 27]
```diff
@@ -837,15 +843,33 @@ def get_node_weight(node) -> int:
...
             return mem_sz * 2
 
     nx_graph = nx.DiGraph()
-    for node in full_bw_graph.nodes:
+    banned_nodes = set()
+
+    def ban_recomputation_if_allowed(node):
+        # This bans recomputation of the node unless we've been forced not to by
+        # user annotation
+        # NB: "recompute" == 0 means it's not banned from recomputing
```
Comment (anijain2305) [inline torch/_functorch/partitioners.py:857]: Not sure about this. We marked the output nodes of AC blocks with "recompute" == 0. Shouldn't we ban recomputation for those nodes?

Comment (Chillee) [reply torch/_functorch/partitioners.py:857]: This was correct (in some sense), forward fix here: https://github.com/pytorch/pytorch/pull/122841
